### ISSUE TITLE
Nan Loss while using resnet_fpn(not pretrained) backbone with FasterRCNN

### ISSUE THREAD

**Issue by sahilg06** [pytorch/vision#4147]
## 🐛 Bug
I am using this model
```
from torchvision.models.detection import FasterRCNN
from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
backbone = resnet_fpn_backbone(backbone_name='resnet152', pretrained=False)
model = FasterRCNN(backbone,
                       num_classes=2)
```
When I set pretrained=True in the backbone, it works absolutely fine but when I set pretrained=False. It starts giving such output

```
Epoch: [0]  [  0/457]  eta: 0:27:14  lr: 0.000003  loss: 141610976.0000 (141610976.0000)  loss_classifier: 50311224.0000 (50311224.0000)  loss_box_reg: 62420652.0000 (62420652.0000)  loss_objectness: 7461720.0000 (7461720.0000)  loss_rpn_box_reg: 21417388.0000 (21417388.0000)  time: 3.5773  data: 0.8030  max mem: 10427
Loss is nan, stopping training
{'loss_classifier': tensor(nan, device='cuda:1', grad_fn=<NllLossBackward>), 'loss_box_reg': tensor(nan, device='cuda:1', grad_fn=<DivBackward0>), 'loss_objectness': tensor(nan, device='cuda:1', grad_fn=<BinaryCrossEntropyWithLogitsBackward>), 'loss_rpn_box_reg': tensor(nan, device='cuda:1', grad_fn=<DivBackward0>)}
```

## Environment

```
torch-version =  1.9.0a0+c3d40fd
torchvision-version =  0.10.0a0
```

Using this dockerfile
```
FROM nvcr.io/nvidia/pytorch:21.06-py3

RUN pip install   pytorch-lightning
RUN pip install -U git+https://github.com/albu/albumentations --no-cache-dir
RUN pip install --upgrade albumentations 
RUN pip install timm
RUN pip install odach
RUN pip install ensemble_boxes
RUN pip install opencv-python-headless
RUN pip install --no-cache-dir --upgrade pip

RUN apt update && apt install -y libsm6 libxext6
RUN apt-get install -y libxrender-dev

RUN apt install -y p7zip-full p7zip-rar
```
Help!

cc @datumbox

**Comment by NicolasHug** [comment]
@sahilg06 could you please provide a minimal reproducing example with code and data? It is very hard to help otherwise, as we don't know the code you're running.

In this specific case it seems your loss is arbitrarily large, it can potentially be caused by the input being out of the expected range

**Comment by sahilg06** [comment]
@NicolasHug Thanks for replying. Does the above model expect a different kind of inputs in both cases? (when the backbone is pretrained and when the backbone is not pretrained). Because just passing pretrained=True is working fine with the rest of the code same.

**Comment by NicolasHug** [comment]
In theory no, but it's possible that your input is wrong in both cases and that the pre-trained model is a bit more well-behaved w.r.t. bad input. Either way, it's hard to help without the full (and minimal) code

**Comment by sahilg06** [comment]
I am using this dataset https://www.aicrowd.com/challenges/global-wheat-challenge-2021/dataset_files.

Here is my dataset class
```
class WheatDataset_training(Dataset):
    
    def __init__(self, df, base_dir):
        self.df = df
        self.base_dir =  base_dir
        self.image_name = df['image_name']
        self.transform = transforms.Compose(
            [
                transforms.ToTensor(),
            ]
        )
    
    def __len__(self):
        return len(self.image_name)
    
    def __getitem__(self, index):
        image_name = self.df.iloc[index].image_name

        path = os.path.join(self.base_dir, 'train', 'train', f'{image_name}.png')

        #for trainng with pseudo-labels
        if not os.path.isfile(path):
            path = os.path.join(self.base_dir, 'test', f'{image_name}.png')

        image = Image.open(path)
        bboxes_str = self.df.iloc[index].BoxesString
        bboxes = bboxes_str.split(';')
        n_objects = len(bboxes)  # Number of wheat heads in the given image
        
        boxes, areas = [], []
        for bbox in bboxes:
          if(bbox!='no_box'):
            box = list(map(float,bbox.split()))
            area = my_utils.area(box)
            if(area > 200000):
                continue
            boxes.append(box)
            areas.append(area)
          else:
            n_objects = 0
        
        # Convert everything into a torch.Tensor
        boxes = torch.as_tensor(boxes, dtype=torch.float32)
        
        # Get the labels. We have only one class (wheat head)
        labels = torch.ones((n_objects, ), dtype=torch.int64)
        
        areas = torch.as_tensor(areas)
        
        # suppose all instances are not crowd
        iscrowd = torch.zeros((n_objects, ), dtype=torch.int64) 
        
        if(n_objects == 0):
          boxes = torch.zeros((0,4), dtype=torch.float32)
          labels =  torch.zeros(0, dtype=torch.int64)
          areas = torch.zeros(0, dtype=torch.float32)
          iscrowd = torch.zeros((0,), dtype=torch.int64)

        target = {
            'boxes': boxes,
            'labels': labels,
            'image_id': torch.tensor([index], dtype=torch.int64), 
            'area': areas,
            'iscrowd': iscrowd
        }
        image = self.transform(image)
        target['boxes'] = target['boxes'].float()
        return image, target
 ```

**Training:**

```
import os
import sys
import numpy as np 
import pandas as pd

import torch
print('torch-version = ',torch.__version__ )
from PIL import Image

from torch import nn
import torchvision
print('torchvision-version = ',torchvision.__version__)
from torch.utils.data import DataLoader
from torchvision import transforms

import models
import my_utils
from datasets import WheatDataset_training

# To select the gpu
torch.cuda.set_device(1)

base_dir = "/raid/sahil_g_ma/wheatDetection"

sys.path.append(os.path.join(base_dir, 'detection'))
from engine import train_one_epoch, evaluate
import utils

train_df = pd.read_csv(os.path.join(base_dir, 'train', 'train.csv'))
# images at index 7 and 72 are same , similarly at 16 and 85 are same
# Also they have inappropriate labels
# hence dropping them to avoid key error in dataloader and to ensure better training
train_df = train_df.drop([7, 72, 16, 85], axis=0)

# To avoid training on images with no_box

#train_df = train_df[train_df.BoxesString != 'no_box']
train_df = train_df.reset_index(drop=True)

# train on the GPU or on the CPU, if a GPU is not available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")


# get the model using our helper function
model = models.FRCNN_resnetfpn_backbone(pre_trained=False)

# move model to the right device
model.to(device)

# use our dataset and defined transformations
dataset = WheatDataset_training(train_df, base_dir)
dataset_test = WheatDataset_training(train_df, base_dir)

# split the dataset in train and test set
#debugging
#indices = [i for i in range(len(dataset))]
indices = torch.randperm(len(dataset)).tolist()
dataset_train = torch.utils.data.Subset(dataset, indices[:-100])
dataset_validation = torch.utils.data.Subset(dataset_test, indices[-100:])

data_loader = torch.utils.data.DataLoader(
    dataset, batch_size=8, shuffle=True, num_workers=2,
    collate_fn=utils.collate_fn)

data_loader_validation = torch.utils.data.DataLoader(
    dataset_validation, batch_size=8, shuffle=False, num_workers=2,
    collate_fn=utils.collate_fn)

# construct an optimizer
params = [p for p in model.parameters() if p.requires_grad]
optimizer = torch.optim.SGD(params, lr=0.01, 
                            momentum=0.9, weight_decay=0.0005)

# and a learning rate scheduler
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer,
                                               step_size=6,
                                               gamma=0.1)

# let's train it for 15 epochs
num_epochs = 25

for epoch in range(num_epochs):
    # train for one epoch, printing every 10 iterations
    train_one_epoch(model, optimizer, data_loader, device, epoch, print_freq=100)
    # update the learning rate
    lr_scheduler.step()
    # evaluate on the test dataset
    evaluate(model, data_loader_validation, device=device)
    print('done')

```

**Comment by NicolasHug** [comment]
Could you please make the example a bit more minimal? For example, do you observe the same issue with randomly-generated input?

**Comment by oke-aditya** [comment]
Also I think 
```
# get the model using our helper function
model = models.FRCNN_resnetfpn_backbone(pre_trained=False)

```

You need to set `num_classes = 2` as well for the model.

P.S. Is this kaggle wheat detection challenge? I had trained faster rcnn on this and it seemed to work fine.

**Comment by sahilg06** [comment]
Thanks, @oke-aditya.  I am specifying the num_classes. Actually, the model I am using above is this (which is in some models.py file) :

```
def FRCNN_resnetfpn_backbone(backbone_name='resnet101', pre_trained=True):

    # Reference: https://github.com/pytorch/vision/blob/master/torchvision/models/detection/backbone_utils.py

    backbone = resnet_fpn_backbone(backbone_name, pretrained=pre_trained)
    
    model = FasterRCNN(backbone,
                       num_classes=2)
    return model
```
and also the format is the same as (xmin, ymin, xmax, ymax) (also I got a decent score by using pretrained backbone(0.704 avg domain accuracy), but now I was trying to train from scratch which resulted in nan loss)

Yes this is 2nd version of that wheat detection challenge (https://www.aicrowd.com/challenges/global-wheat-challenge-2021)

**Comment by sahilg06** [comment]
Hi @oke-aditya, I recently realized that the format was same (xmin, ymin, xmax, ymax) but in the given data they have taken bottom left corner as origin, whereas in pascal-voc, top left is taken as origin. Does this matter?

**Comment by sahilg06** [comment]
Hi @NicolasHug, Sorry for the very late reply. Is this okay as a minimal code with randomly generated input?
```
import torch
from torchvision.models.detection import FasterRCNN
from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

backbone = resnet_fpn_backbone(backbone_name='resnet152', pretrained=False)
model = FasterRCNN(backbone,
                       num_classes=2)
model.to(device)
image = torch.rand(1,3,1024,1024)
boxes = torch.zeros((0,4), dtype=torch.float32)
labels =  torch.zeros(0, dtype=torch.int64)
areas = torch.zeros(0, dtype=torch.float32)
iscrowd = torch.zeros((0,), dtype=torch.int64)

target = {
    'boxes': boxes.to(device),
    'labels': labels.to(device),
    'image_id': torch.tensor([0], dtype=torch.int64).to(device), 
    'area': areas.to(device),
    'iscrowd': iscrowd.to(device)
}

params = [p for p in model.parameters() if p.requires_grad]
optimizer = torch.optim.SGD(params, lr=0.01, 
                            momentum=0.9, weight_decay=0.0005)
for i in range(4):
  loss_dict = model(image.to(device), [target])
  losses = sum(loss for loss in loss_dict.values())
  print(losses)
  optimizer.zero_grad()
  losses.backward()
  optimizer.step()
```

when I set `pretrained=False` in the backbone, I got the following output
```
tensor(nan, device='cuda:0', grad_fn=<AddBackward0>)
tensor(nan, device='cuda:0', grad_fn=<AddBackward0>)
tensor(nan, device='cuda:0', grad_fn=<AddBackward0>)
tensor(nan, device='cuda:0', grad_fn=<AddBackward0>)
```

and when I set 'pretrained=True'  in the backbone, I got the following output

```
tensor(1.3928, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.8798, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.7465, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.2988, device='cuda:0', grad_fn=<AddBackward0>)
```

**Comment by datumbox** [comment]
@sahilg06 Thanks for providing such a good example and apologies for the late response. I was able to reliably reproduce the problem.

The NaNs are the source of two separate issues:

### 1. Empty RPN proposals

Because the network is so deep and you initialize it with random weights, it happens that the initial proposals of RPN are of very poor quality. As a result the post processing/filtering steps removes them all. This happens at:
https://github.com/pytorch/vision/blob/c8f7d772e844d707e152e2a1fa1aad26cf1b7530/torchvision/models/detection/rpn.py#L355-L356

Above the `proposals` are none empty but the `boxes` are empty. This is because of the small box removal that happens at:
https://github.com/pytorch/vision/blob/c8f7d772e844d707e152e2a1fa1aad26cf1b7530/torchvision/models/detection/rpn.py#L262

The empty proposals eventually lead to calling `fastrcnn_loss()` with empty predictions and targets which causes the NaNs.

To ensure your model has sufficient proposals you can try setting `model.rpn.min_size = 0.0` after you initialize the model. Note that you might need to tweak some of the other filtering parameters as well.

### 2. Too high LR, too low weight decay

By fixing the above the first loss call will be non NaN. But the subsequent runs can still lead to NaN values. This is because of the massive loss value which leads to overflows on the weight param estimation.

You can fix the issue by reconfiguring your optimizer:
```python
optimizer = torch.optim.SGD(params, lr=1e-8, momentum=0.9, weight_decay=0.05)
```


Fixing the above should allow you to consistently execute your snippet without hitting nans:
```
tensor(5870150.5000, grad_fn=<AddBackward0>)
tensor(7591120.5000, grad_fn=<AddBackward0>)
tensor(0., grad_fn=<AddBackward0>)
tensor(0., grad_fn=<AddBackward0>)
```

Finally please note that as far as I see the problem is reproducible only with very deep networks such as `resnet152`. Using shallower networks like `resnet50` usually avoids problem 1 (not problem 2 though, still you need to adapt your optimizer config). This intuitively makes sense because very deep networks initialized randomly are more likely to yield low quality initial proposals. For most applications it's worth pre-training the backbone on a separate dataset or starting with some pre-trained weights.

I believe the above answers the question so I'm going to close the issue to keep things tidy. If you have more questions, feel free to reopen it.
